keps/sig-apps/0003-kubernetes-application-api.md

```diff
@@ -0,0 +1,1120 @@
+              key: password
+        ports:
+        - containerPort: 80
+          name: wordpress
+        volumeMounts:
+        - name: wordpress-persistent-storage
+          mountPath: /var/www/html
+  volumeClaimTemplates:
+       - metadata:
+           name: mysql-persistent-storage
+         spec:
+           accessModes: [ "ReadWriteOnce" ]
+           resources:
+             requests:
+               storage: 250Gi
+```
+
+Above we generate two Applications, one for the MySQL RDBMs and another for the WordPress web server. Each of these 
+Applications can be managed independently. For instance, if the WordPress web server is deleted, the components of the 
+MySQL RDBMS will not be affected. However, the dependency between the two workloads is made explicit in the 
+`Spec.Dependencies` of the `wordpress` Application. Tools and UIs that examine the `wordpress` application are made 
+aware of its dependency on the `mysql` Application, and they can surface the dependency to users.
+
+#### Selecting Components
+In order to find the `Spec.Componenets` of an Application, a client should perform the following.
+
+1. For each GroupVersionKind in the Applications `Spec.Components`, list all objects of that Kind, in the namespace of 
+the Application, that match the Application's `Spec.Selector`.
+1. For each selected object, validate that the object has a `kubernetes.io/application` that matches the name of the 
+Application. Discard any objects that do not have such an annotation. This ensures that selector overlap does not 
+violate the intentions of the application creators and administrators.
+1. Unless otherwise indicated, clients should ensure that the `OwnerReferences` of the object contain the `UID` of the 
+Application.
+
+#### Retrieving Applications
+In order to retrieve the Application corresponding to an API object, a client should perform the following.
+
+1. Examine the object for a `kubernetes.io/application` annotation.
+1. If no annotation is found, the object is not associated with an Application.
+1. If the annotation is found, retrieve the Application object with the indicated name from the same namespace of the 
+object.
+1. If no such Application is found, the Application is either deleted or has not yet been created.
+1. If the Application is found, validate that the `Spec.Selector` of the Application object matches the labels of the 
+object. If the object's labels do not match, the object has been orphaned.
+1. Check the `OwnerReferences` of the object for a reference to the Application by `UID` to ensure that ownership has 
+been established.
+
+#### Retrieving Dependencies
+In order to retrieve the `Spec.Dependencies` of an Application, a client should perform the following.
+
+1. For each name in the `Spec.Dependencies` list, retrieve the Application of that name, in the namespace of the 
+Application under consideration. If no such Application is found, the dependency is not installed.
```

Comment:
This seems to provide a bit of a UX conundrum. Say I install an application with a dependency on `foo`. And say I don't have a `foo` already in my cluster. Given that the name itself is not guaranteed to be semantically meaningful (as the name `foo` certainly is not), I have absolutely no way of knowing what I am actually missing. Furthermore, I can "satisfy" the dependency by creating an empty Application named `foo`, but in no way is that actually _satisfying_ the dependency that the application actually has.

So again, I am not understanding what is gained by such a weak notion of "dependency". It certainly doesn't increase the discoverability of the dependency, since the string name's only semantic property is uniqueness within a (presumably) namespace.

Reply:
This should probably also state explicitly how it will solve or avoid the circular dependency problem and the diamond dependency problem.

The circular dependency problem is a valid problem case (we can easily imagine a DB system dependig on a monitoring system which depends on a DB system). But since applications don't seem to have any notion of readiness, it's probably easy enough to solve just by installing all related apps at the same time. But that solution may have operational issues.

The diamond dependency problem is perhaps slightly worsened by the fact that this proposal uses only a string (again, one with no semantic properties other than local uniqueness) to identify a dependency. So there is no path to automatic dependency resolution available in the current proposal.

Reply:
This doesn't deal with the packaging problem.